docs/architecture/adr-049-state-sync-hooks.md

```diff
@@ -0,0 +1,187 @@
+implement `ExtensionSnapshotter` as extension snapshotters. When setting up the application, the snapshot `Manager` should call 
+`RegisterExtensions([]ExtensionSnapshotter…)` to register all the extension snapshotters.
+
+```proto
+// SnapshotItem is an item contained in a rootmulti.Store snapshot.
+message SnapshotItem {
+  // item is the specific type of snapshot item.
+  oneof item {
+    SnapshotStoreItem        store             = 1;
+    SnapshotIAVLItem         iavl              = 2 [(gogoproto.customname) = "IAVL"];
+    SnapshotExtensionMeta    extension         = 3;
+    SnapshotExtensionPayload extension_payload = 4;
+  }
+}
+
+// SnapshotStoreItem contains metadata about a snapshotted store.
+message SnapshotStoreItem {
+  string name = 1;
+}
+
+// SnapshotIAVLItem is an exported IAVL node.
+message SnapshotIAVLItem {
+  bytes key     = 1;
+  bytes value   = 2;
+  int64 version = 3;
+  int32 height  = 4;
+}
+
+// SnapshotExtensionMeta contains metadata about an external snapshotter.
+// One module may need multiple snapshotters, so each module may have multiple SnapshotExtensionMeta.
+message SnapshotExtensionMeta {
+  string name   = 1;
+  // format is used within the snapshotter/namespace, not global one for all modules
+  uint32 format = 2;
+}
+
+// SnapshotExtensionPayload contains payloads of an external snapshotter.
+message SnapshotExtensionPayload {
+  bytes payload = 1;
+}
+```
+
+The snapshot stream would look like this:
+
+```go
+// multi-store snapshot
+{SnapshotStoreItem | SnapshotIAVLItem, ...}
+// extension1 snapshot
+SnapshotExtensionMeta
+{SnapshotExtensionPayload, ...}
+// extension2 snapshot
+SnapshotExtensionMeta
+{SnapshotExtensionPayload, ...}
+```
+
+Add `extensions` field to snapshot `Manager` for extension snapshotters. `multistore` snapshotter is a special one and it doesn't need a name because it is always placed at the beginning of the binary stream.
+
+```go
+type Manager struct {
+	store      *Store
+	multistore types.Snapshotter
+	extensions map[string]types.ExtensionSnapshotter
+    mtx                sync.Mutex
+	operation          operation
+	chRestore          chan<- io.ReadCloser
+	chRestoreDone      <-chan restoreDone
+	restoreChunkHashes [][]byte
+	restoreChunkIndex  uint32
+}
+```
+
+For extension snapshotters that implement the `ExtensionSnapshotter` interface, their names should be registered to the snapshot `Manager` by 
+calling `RegisterExtensions` when setting up the application. And the snapshotters will handle both taking snapshot and restoration.
+
+```go
+// RegisterExtensions register extension snapshotters to manager
+func (m *Manager) RegisterExtensions(extensions ...types.ExtensionSnapshotter) error 
+```
+
+Remain `Snapshotter` interface for `multistore` snapshotter. No need for `SnapshotFormat` and `SupportedFormats` as for `multistore` snapshotter
+formats are handled in a higher level.
```

Comment:
I don't quite follow this. Consider rephrasing this. What exactly are you trying to describe here?

Reply:
I have removed this part as the `Snapshotter` interface doesn't change